### ISSUE TITLE
[backend-compiler] 3-D expert-weight slab gather (expert_w[ids], [E,H,F]→[M,H,F]) overflows an immediate offset

### ISSUE THREAD

**Issue by ani300** [torch-spyre/torch-spyre#3637]
# [backend-compiler] 3-D expert-weight slab gather (expert_w[ids], [E,H,F]→[M,H,F]) overflows an immediate offset

## Summary

MoE needs to fetch, for each routed row, the weight slab of the expert it was routed
to: gathering rows of a `[E, ...]` expert-weight table with an `[M]` index vector.

- The **2-D** row gather `W[ids]` (`[E,H]`, `[M]` → `[M,H]`) now **works** on device
  and matches CPU.
- The **3-D** slab gather `W3[ids]` (`[E,H,F]`, `[M]` → `[M,H,F]`) — the shape MoE
  gate/up/down weights actually need — **hard-fails** in the backend compiler with an
  immediate-offset overflow.

## Repro

`repros/gemma4_moe/issues/p4_expert_weight_gather.py` (no model load, no real weights):

```python
import torch, torch_spyre
E, H, F, M = 128, 2816, 704, 256   # Gemma-4 26B-A4B: 128 experts, moe_inter=704

def row_gather(W, ids):  return W[ids]    # [E,H],[M]   -> [M,H]
def slab_gather(W3, ids): return W3[ids]  # [E,H,F],[M] -> [M,H,F]

ids = torch.randint(0, E, (M,), dtype=torch.int64)
torch.compile(row_gather,  dynamic=False)(torch.randn(E,H).half().to("spyre"),      ids.to("spyre")).cpu()
torch.compile(slab_gather, dynamic=False)(torch.randn(E,H,F).half().to("spyre"),    ids.to("spyre")).cpu()
```

## Observed (on-device, 2026-08-10)

```
row [E,H]    : OK (matches CPU)
slab [E,H,F] : terminate called after throwing an instance of 'DtException'
   what():  DtException: Program verification failed for core 0 node 1_identity
   Error message: Immediate value out of boundary in instruction:
   L3_ADDEARIMM :: be:be imm:-7929824 src0:0
   , file .../deeptools/dxp/dxp.cpp line 133
```

The 3-D gather emits an address-computation immediate (`-7929824`) that overflows the
`L3_ADDEARIMM` immediate range — the per-row stride of an `[E, H=2816, F=704]` slab
(H*F elements) exceeds what the addressing instruction can encode.

## Impact

MoE expert weights are 3-D (`[E, H, F]` for gate/up, `[E, F, H]` for down). Selecting
per-row expert weights by gather is not usable at these sizes. Note the corresponding
inductor tests are currently guarded: `tests/inductor/indirect_access_common.py:413-434`
(xfail) and `tests/inductor/test_indirect_access_gather.py:447-465` (`test_moe` skipped).

## Suggested fix / discussion

- Backend addressing for indirect gather should handle per-row strides beyond the
  `L3_ADDEARIMM` immediate range (e.g. materialize the base offset in a register
  rather than as an immediate), or
- lower a 3-D slab gather as a loop of 2-D gathers, which are known-good here.

Either way the 2-D vs 3-D split is the actionable boundary: 2-D gather is fine, 3-D
overflows.

## Labels

bug, backend-compiler

**Comment by ani300** [comment]
This now works as expected
